main.cpp

```diff
@@ -24,328 +24,243 @@ llvm::cl::extrahelp ConstexprCategoryHelp(R"(
-                                       "function can be constexpr");
+    const auto ID = DE.getCustomDiagID(clang::DiagnosticsEngine::Warning, "function can be constexpr");
 
     DE.Report(loc, ID).AddFixItHint(FixIt);
 
     return true;
   }
 };
 
-class ConstexprVarDeclFunctionASTVisitor
-    : public clang::RecursiveASTVisitor<ConstexprVarDeclFunctionASTVisitor> {
-  clang::SourceManager &sourceManager_;
-  clang::CompilerInstance &CI_;
-  clang::DiagnosticsEngine &DE;
+class ConstexprVarDeclFunctionASTVisitor : public clang::RecursiveASTVisitor<ConstexprVarDeclFunctionASTVisitor>
+{
+  clang::SourceManager& sourceManager_;
+  clang::CompilerInstance& CI_;
+  clang::DiagnosticsEngine& DE;
 
-  class ConstexprVarDeclVisitor
-      : public clang::RecursiveASTVisitor<ConstexprVarDeclVisitor> {
-    clang::CompilerInstance &CI_;
-    clang::DiagnosticsEngine &DE;
+  class ConstexprVarDeclVisitor : public clang::RecursiveASTVisitor<ConstexprVarDeclVisitor>
+  {
+    clang::CompilerInstance& CI_;
+    clang::DiagnosticsEngine& DE;
 
-  public:
-    explicit ConstexprVarDeclVisitor(clang::CompilerInstance &ci)
-        : CI_(ci), DE(ci.getASTContext().getDiagnostics()) {}
+   public:
+    explicit ConstexprVarDeclVisitor(clang::CompilerInstance& ci) : CI_(ci), DE(ci.getASTContext().getDiagnostics()) {}
 
-    bool VisitDeclStmt(clang::DeclStmt *stmt) {
-      if (!stmt->isSingleDecl())
+    bool VisitDeclStmt(clang::DeclStmt* stmt) {
+      if (!stmt->isSingleDecl()) {
         return true;
+      }
 
-      clang::VarDecl *var =
-          clang::dyn_cast<clang::VarDecl>(*stmt->decl_begin());
-      if (!var)
+      auto* var = clang::dyn_cast<clang::VarDecl>(*stmt->decl_begin());
+      if (!var) {
         return true;
+      }
 
       // Skip variables that are already constexpr
-      if (var->isConstexpr())
+      if (var->isConstexpr()) {
         return true;
+      }
 
       // Only do locals for right now
-      if (!var->hasLocalStorage())
+      if (!var->hasLocalStorage()) {
         return true;
+      }
 
-      clang::SourceLocation loc = stmt->getSourceRange().getBegin();
-      auto &sema = CI_.getSema();
+      clang::SourceLocation const loc = stmt->getSourceRange().getBegin();
+      // auto& sema = CI_.getSema();
 
       // var needs an initializer
-      Expr *Init = var->getInit();
-      if (!Init)
+      Expr* Init = var->getInit();
+      if (!Init) {
         return true;
+      }
 
       // If the var is const we can mark it constexpr
-      QualType ty = var->getType();
-      if (!ty.isConstQualified())
+      QualType const ty = var->getType();
+      if (!ty.isConstQualified()) {
         return true;
+      }
 
       // Is init an integral constant expression
-      if (!var->checkInitIsICE())
+#if LLVM_VERSION_MAJOR >= 12
+      if (!var->hasICEInitializer(CI_.getASTContext())) {
+        //if (!var->hasICEInitializer(stmt->getSingleDecl()->getASTContext())) {
```

Comment:
Not sure which of the two to use.

But added at 12 https://github.com/llvm/llvm-project/commit/3692d20d2b994ce865ffb97096b05218279e1ebd

Reply:
This looks correct -- they should boil down to the same `ASTContext`.